### PR TITLE
refactor: Use TensorViewer in SciPy optimizer, instead of explicit constraints

### PR DESCRIPTION
**Pull request by kratsg**
# Pull Request Description

Part of pulling out/refactoring #951. We need to split out the portion of rewriting scipy to use TensorViewer first, so we know the changes in the tests, before doing a larger refactor.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Rewrite the scipy_optimizer to use TensorViewer and drop constraints
```

### CODE REVIEW

File: src/pyhf/__init__.py [hunk 1]
```diff
@@ -4,10 +4,8 @@
 from .exceptions import InvalidBackend
 from . import events
 
-tensorlib = tensor.numpy_backend()
-default_backend = tensorlib
-optimizer = optimize.scipy_optimizer()
-default_optimizer = optimizer
+tensorlib = None
+optimizer = None
```
Comment (matthewfeickert) [inline src/pyhf/__init__.py:8]: Why do we need to split these up into here and the section below?

File: src/pyhf/optimize/opt_scipy.py [hunk 2]
```diff
@@ -28,26 +30,50 @@ def minimize(
...
+        variable_idx = [x for x in all_idx if x not in fixed_idx]
+        variable_init = all_init[variable_idx]
+        variable_bounds = [par_bounds[i] for i in variable_idx]
+
+        tv = _TensorViewer([fixed_idx, variable_idx])
+
+        data = tensorlib.astensor(data)
+        fixed_values_tensor = tensorlib.astensor(fixed_values, dtype='float')
+
+        def func(pars):
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_scipy.py:53]: Maybe 
```suggestion
        def _objective(pars):
```
to keep it clear?

File: src/pyhf/optimize/opt_scipy.py [hunk 3]
```diff
@@ -28,26 +30,50 @@ def minimize(
...
+            pars = tensorlib.astensor(pars)
+            constrained_pars = tv.stitch([fixed_values_tensor, pars])
+            return objective(constrained_pars, data, pdf)
+
         result = minimize(
-            objective,
-            init_pars,
-            constraints=constraints,
+            func,
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_scipy.py:59]: Maybe
```suggestion
            _objective,
```
to keep it clear?

File: tests/test_validation.py [hunk 4]
```diff
@@ -709,13 +709,13 @@ def validate_hypotest(pdf, data, mu_test, expected_result, tolerance=1e-6):
...
-        (setup_1bin_shapesys(), 1e-6),
-        (setup_1bin_lumi(), 4e-6),
-        (setup_1bin_normsys(), 2e-9),
-        (setup_2bin_histosys(), 8e-5),
-        (setup_2bin_2channel(), 1e-6),
-        (setup_2bin_2channel_couplednorm(), 1e-6),
-        (setup_2bin_2channel_coupledhistosys(), 1e-6),
+        (setup_1bin_shapesys(), 4e-5),
+        (setup_1bin_lumi(), 6e-5),
+        (setup_1bin_normsys(), 2e-4),
```
Comment (matthewfeickert) [inline tests/test_validation.py:714]: Do we know why normsys seems to be so sensitive here? This tolerance is much larger than all the others.

File: src/pyhf/__init__.py (hunk 1 above)
Comment (kratsg) [reply src/pyhf/__init__.py:8]: cyclic import. the default optimizer (scipy) needs the `get_backend` function call which depends on those variables being global, but they can't be set until after `get_backend` is defined.

File: src/pyhf/optimize/opt_scipy.py (hunk 2 above)
Comment (kratsg) [reply src/pyhf/optimize/opt_scipy.py:53]: `func` is in all the other backends right now, so we're going for consistency. We can rename in the next PR that completely removes all of these changes.

File: src/pyhf/optimize/opt_scipy.py (hunk 3 above)
Comment (kratsg) [reply src/pyhf/optimize/opt_scipy.py:59]: func is in all the other backends right now, so we're going for consistency. We can rename in the next PR that completely removes all of these changes.

File: tests/test_validation.py (hunk 4 above)
Comment (kratsg) [reply tests/test_validation.py:714]: Absolutely no idea. Fitted parameters are the same though down to numerical precision.

Comment (lukasheinrich) [reply tests/test_validation.py:714]: for fixed POI fits I would expect the fitted paramerters to differ

Comment (kratsg) [reply tests/test_validation.py:714]: ahhhhh, yes, I never checked the fixed poi fits...

Comment (kratsg) [reply tests/test_validation.py:714]: for 1bin_normsys, we have

master
```
# unconstrained
(Pdb) pyhf.infer.mle.fit(data, pdf).tolist()
[0.30359961840184535, 0.34708322040569417]
# constrained
(Pdb) pyhf.infer.mle.fixed_poi_fit(1.0, data, pdf).tolist()
[1.0, -1.7880486985341282]
```

this branch
```
(Pdb) pyhf.infer.mle.fit(data, pdf).tolist()
[0.30359961840184535, 0.34708322040569417]
(Pdb) pyhf.infer.mle.fixed_poi_fit(1.0, data, pdf).tolist()
[1.0, -1.7880486985341286]
```

Comment (kratsg) [reply tests/test_validation.py:714]: and the CLs results

master
```
#CLs obs
(Pdb) CLs_obs.tolist()
[0.0006735317010546278]
(Pdb) expected_result['obs']
0.0006735317023683173
(Pdb) (abs(CLs_obs - expected_result['obs']) / expected_result['obs']).tolist()
[1.950449393019978e-09]
# CLs expected
(Pdb) CLs_exp_set.ravel().tolist()
[7.471694617552582e-10, 5.741155155228773e-08, 3.6898088058366224e-06, 0.00016965731536561725, 0.004392708998221363]
(Pdb) expected_result['exp']
[7.47169462e-10, 5.7411551509088054e-08, 3.6898088062731205e-06, 0.00016965731538267896, 0.004392708998555453]
(Pdb) ((CLs_exp_set.ravel() - expected_result['exp'])/expected_result['exp']).tolist()
[-3.275585606661175e-10, 7.52456113642322e-10, -1.1829829768812686e-10, -1.0056572025296775e-10, -7.60556627694298e-11]
```

this branch
```
# CLs obs
(Pdb) CLs_obs.tolist()
[0.000673533629944247]
(Pdb) expected_result['obs']
0.0006735317023683173
(Pdb) (abs(CLs_obs - expected_result['obs']) / expected_result['obs']).tolist()
[2.8618933940777534e-06]

# CLs expected
(Pdb) CLs_exp_set.ravel().tolist()
[7.4725818074954e-10, 5.741738540786033e-08, 3.69012109380142e-06, 0.00016966883322846522, 0.0043929353930099325]
(Pdb) expected_result['exp']
[7.47169462e-10, 5.7411551509088054e-08, 3.6898088062731205e-06, 0.00016965731538267896, 0.004392708998555453]
(Pdb) ((CLs_exp_set.ravel() - expected_result['exp'])/expected_result['exp']).tolist()
[0.0001187397960598631, 0.00010161541743657144, 8.463515176413025e-05, 6.788888389678545e-05, 5.153868707301546e-05]
```

Comment (kratsg) [reply tests/test_validation.py:717]: To wrap up discussion here --- two tests were added to compare the fitted parameter differences that come out of the minimization routines (optimizers) and ensure they don't change within tolerance. It seems that the CLs has a lot more computations that tiny changes in fitted parameters inflates to larger differences in CLs.

**Comment by kratsg** [comment]
Closing in favor of #951 which is preserving all of the functionality instead of removing the option to not-stitch entirely.

**Pull request closed without merging**